### PR TITLE
Better implementation for micromega proof compaction

### PR DESCRIPTION
**Pull request by ppedrot**
This is both cleaner and more efficient, though still quite naive. After this PR, the runtime of the Nia.v test file went from ~35s down to ~25s on my machine, and the profile now essentially only displays VM computations, rather than hypothesis selection as it was before.

### CODE REVIEW

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1954425847

**Comment by Alizter** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                             │                         │                                       │                                       │                         │                   │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD    PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼───────────────────┤
│                    coq-core │  100.41   101.57  -1.14 │   381437549687    380107714907   0.35 │   433071890368    433162135254  -0.02 │  249572   248840   0.29 │   0    0      nan │
│                   coq-verdi │   49.58    49.96  -0.76 │   216020335958    217535379109  -0.70 │   324093139857    324065762596   0.01 │  829636   832324  -0.32 │   0    0      nan │
│                   coq-color │  270.72   271.87  -0.42 │  1175959043414   1181554246218  -0.47 │  1598962864988   1599428555552  -0.03 │ 1141480  1142996  -0.13 │   1    2   -50.00 │
│            coq-math-classes │   96.55    96.92  -0.38 │   418699480845    420276628324  -0.38 │   591687691352    591677460203   0.00 │  495444   496296  -0.17 │   0    0      nan │
│                coq-rewriter │  384.95   385.79  -0.22 │  1680401670227   1684123144152  -0.22 │  2770765186292   2770497269063   0.01 │ 1064148  1062628   0.14 │  15    4   275.00 │
│                  coq-flocq3 │   78.88    79.05  -0.22 │   343596178257    343985887210  -0.11 │   454915010661    455022184982  -0.02 │  987652   987992  -0.03 │   8    0      nan │
│           coq-iris-examples │  474.49   475.43  -0.20 │  2072097230809   2076683615360  -0.22 │  3139567289064   3139964720424  -0.01 │ 1180852  1180892  -0.00 │   0    0      nan │
│ coq-rewriter-perf-SuperFast │  952.41   953.70  -0.14 │  4162487043395   4169113654162  -0.16 │  7193899722602   7195732746373  -0.03 │ 1144256  1215312  -5.85 │   4    0      nan │
│                 coq-bignums │   29.18    29.21  -0.10 │   127527788822    127542869711  -0.01 │   176080088201    176119934742  -0.02 │  469216   465696   0.76 │   0    0      nan │
│                coq-coqprime │  183.52   183.67  -0.08 │   803684437837    804270352505  -0.07 │  1559913439088   1560090853986  -0.01 │  858704   856196   0.29 │   5    0      nan │
│                    coq-corn │  823.79   824.39  -0.07 │  3597404624532   3599116482858  -0.05 │  5587771087582   5587796588597  -0.00 │  876392   874484   0.22 │   0    0      nan │
│          coq-mathcomp-field │  117.57   117.59  -0.02 │   515243677518    515597185985  -0.07 │   851252992656    851237277516   0.00 │  663820   661148   0.40 │   0    0      nan │
│                 coq-unimath │ 3745.14  3745.59  -0.01 │ 16398538739376  16401864988896  -0.02 │ 32896186249046  32893096953328   0.01 │ 3873680  3873752  -0.00 │   2    3   -33.33 │
│                     coq-vst │ 1185.38  1185.46  -0.01 │  5178005004251   5179475054726  -0.03 │  8413606152940   8413358028277   0.00 │ 1892788  1892916  -0.01 │   4    0      nan │
│             coq-fiat-crypto │ 3042.83  3041.98   0.03 │ 13281993916891  13278226765630   0.03 │ 24365863774964  24372568479306  -0.03 │ 3871080  3871028   0.00 │  86   98   -12.24 │
│            coq-fiat-parsers │  339.18   339.07   0.03 │  1464955218306   1467151418689  -0.15 │  2422101739887   2422216446479  -0.00 │ 2787508  2787480   0.00 │   6   20   -70.00 │
│  coq-performance-tests-lite │  805.51   804.96   0.07 │  3505850534752   3503329218655   0.07 │  5960382962918   5960884124316  -0.01 │ 1827544  1629000  12.19 │   3    0      nan │
│                 coq-coqutil │   35.82    35.79   0.08 │   154617320701    154467962608   0.10 │   208043234653    208265681536  -0.11 │  600708   601368  -0.11 │   0    0      nan │
│                coq-bedrock2 │  194.48   194.31   0.09 │   849397534663    848927974476   0.06 │  1570921287570   1570915945384   0.00 │ 2053988  2053916   0.00 │ 112    6  1766.67 │
│               coq-perennial │ 4239.99  4236.19   0.09 │ 18545532991354  18528285012013   0.09 │ 30690093933604  30697587227134  -0.02 │ 2610392  2604504   0.23 │   0    0      nan │
│       coq-mathcomp-solvable │   90.62    90.53   0.10 │   396377052441    396074486453   0.08 │   612092168025    612131699801  -0.01 │  670796   672312  -0.23 │   0    0      nan │
│      coq-mathcomp-ssreflect │   26.61    26.58   0.11 │   115974249551    115847714016   0.11 │   147486862979    147458556891   0.02 │  531440   532224  -0.15 │   0    0      nan │
│                  coq-stdlib │  416.26   415.79   0.11 │  1499084676479   1488985217659   0.68 │  1445628636817   1445725120002  -0.01 │  589304   589032   0.05 │   0    0      nan │
│               coq-fiat-core │   55.45    55.38   0.13 │   231235702930    231303822207  -0.03 │   324295340433    324284967764   0.00 │  473512   475540  -0.43 │   7    0      nan │
│               coq-fourcolor │ 1519.02  1516.81   0.15 │  6642425021990   6634551492497   0.12 │ 12086581071671  12086878618979  -0.00 │  725684   725620   0.01 │   0    0      nan │
│      coq-mathcomp-character │   78.75    78.63   0.15 │   344996191662    344384013178   0.18 │   527780576468    527784999109  -0.00 │  727472   727460   0.00 │   0    0      nan │
│      coq-mathcomp-odd-order │  577.65   576.74   0.16 │  2532544778181   2528374205934   0.16 │  4309377258410   4309460099177  -0.00 │  931620   933424  -0.19 │   0    0      nan │
│              coq-verdi-raft │  601.73   600.63   0.18 │  2627338496769   2623268570415   0.16 │  4047710205886   4047827015604  -0.00 │ 1220708  1221588  -0.07 │   0    0      nan │
│                coq-compcert │  300.23   299.55   0.23 │  1306143275028   1302357279730   0.29 │  1944320385409   1944267569101   0.00 │ 1097912  1099140  -0.11 │   0    7  -100.00 │
│                    coq-hott │  157.80   157.35   0.29 │   683466775836    682689850946   0.11 │  1075407214956   1075409956856  -0.00 │  656372   659076  -0.41 │   0    0      nan │
│                  coq-geocoq │  740.95   738.46   0.34 │  3236010641767   3224195630850   0.37 │  5157207878546   5157185231668   0.00 │  986340   982760   0.36 │   0    0      nan │
│        coq-mathcomp-algebra │   65.51    65.28   0.35 │   286450020205    285615384672   0.29 │   399055884323    399116872381  -0.02 │  560220   560052   0.03 │   0    0      nan │
│       coq-engine-bench-lite │  181.04   180.39   0.36 │   760025722179    757553852776   0.33 │  1483984682839   1485066067750  -0.07 │ 1173772  1173784  -0.00 │   0    0      nan │
│       coq-mathcomp-fingroup │   24.15    24.01   0.58 │   105522123298    104883677425   0.61 │   155201494460    155219318851  -0.01 │  478152   477940   0.04 │   0    0      nan │
│              coq-coquelicot │   35.61    35.31   0.85 │   152968650421    152788186959   0.12 │   202626399412    202627900592  -0.00 │  749040   748808   0.03 │  98    5  1860.00 │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴───────────────────┘
```

**Comment by ppedrot** [comment]
Didn't seem to have much effect on the bench fauna. There is a single line in fiat-crypto that might reflect this but that's about it...

**Review by Alizter — APPROVED** [review]
In any case I think it is a worthwhile change. Especially since the test-suite runs a bit faster. Do you want to wait for feedback from the lia maintainers?

**Comment by ppedrot** [comment]
The lia maintainers are usually not the most responsive ones from the lot, but I don't mind waiting, this is not urgent either.

**Comment by fajb** [comment]
@ppedrot looks good to me. Though I am always scared because I messed up this part of the code so many times.

**Review by fajb — APPROVED** [review]

**Comment by Alizter** [comment]
merging soon

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**
